README.md
# Screenerx

This nodejs project will create batch screenshots of websites.
It uses the awesome puppeteer library which uses a real browser under the hood, e.g. a Chrome or Firefox.

## Installation

```bash
npm install -g screenerx
# or:
yarn global add screenerx
# or use docker (see below)
```

## Usage

### Use arguments

You can call screenerx with arguments:


#### Examples
```bash
# create a screenshot with default settings
screenerx --url https://example.com

# set width and height
screenerx --url https://example.com --width=1200 --height=800

# create a full page screenshot
screenerx --url https://example.com --width=1200 --height=full

# set output path
screenerx --url https://example.com --width=1200 --height=800 --output-path=screenshots/test

# set output file (if not set, the url will be used)
screenerx --url https://example.com --width=1200 --height=800 --output-path=screenshots/test --output-file=example.jpg

# pass an url via stdin (e.g. from a file or echo)
echo "https://example.com" | xargs -I {} ./screenerx.js --url={}
 
# create screenshots in parallel (using gnu-parallel) with 4 threads: 
cat links.txt | parallel --line-buffer -j 4 screenerx --url {}

# docker example (see below for more information)
cat links.txt | parallel --line-buffer -j 4 docker-compose run --rm screenerx --url {}

# specify a link file, default is links.txt (link file is ignored if you specify an url!)
screenerx --file=links.txt
```

### Create a config file

If you like, you can also create a config file. The config file will be used to generate the screenshots.

create a file called `screenerconf.json`, you can see at `screenerconf.json.example` for an example.
An entry looks basically like this:

```json
{
    "linkFile" : "links.txt",
    "tasks": [
        {
            "name" : "test_landscape",
            "enabled" : true,
            "baseUrl" : "https://www.example.com",
            "linkFile" : "links.txt",
            "browser" : "chrome",
            "width" : "1920",
            "height" : "1080"
        }
    ]
}
```

| prop     | description                                                                                                     |
|----------|-----------------------------------------------------------------------------------------------------------------|
| name     | name of the task, this will be used in directory name                                                           |
| enabled  | should this task run?                                                                                           |
| baseUrl  | if you specify a baseUrl, you can use paths only in your urlfile                                                |
| browser  | chrome, firefox, etc.. Default is chrome, if you need another browsers, see puppeteer docs how to install these |
| width    | width of the screen                                                                                             |
| height   | height of the screen, can be also "full" to create full size screenshots                                        |
| linkFile | set a different link file for this job                                                                          |

create a text file and put your links into that file. For example `links.txt` (as specified above in json)
The link list can be a normal list with full URLs starting with http://, https:// or you can also only specify the paths. for example /about. The full URL will be generated out of provided baseUrl in `screenerx.conf.json` and the provided path in links.txt
You can specify the linkFile in the root of the json file, but you can also specify / override it in the task itself.

The screenshots will be made in the screenshot folder.


## Docker

You can also use this tool in a docker container.
Just build the container with `docker build -t screenerx .` and run it with `docker run -it --rm -v $(pwd):/app/screenshots screenerx`

or use the docker-compose file:

```bash
docker-compose run --rm screenerx
```

## License
MIT

## Author
Michael Milawski


CHANGELOG.md
### 1.1.1 - 2023-05-21
Added
- Support for Docker. Added Dockerfile and docker-compose.yml. Puppeteer will be started with --no-sandbox and --disable-setuid-sandbox to make it work in docker.

### 1.1.0 - 2023-05-20
Added
- New arguments: --url, --width, --height, --outputPath, --outputFile have been added.
- Added support for stdin input, so you can for example use gnu-parallel to create screenshots in parallel. See README.md for an example.
- Removed fullPage argument, use --height=full instead.

### 1.0.1 - 2023-04-20
Release
- Initial release.

screenerx.js
#!/usr/bin/env node
const yargs = require('yargs');
const puppeteer = require('puppeteer');
const fs = require('fs').promises;

const screenshotBasePath = 'screenshots';
const screenshotPath = `${screenshotBasePath}/${new Date().toISOString()}`;
const confFile = 'screenerx.conf.json';

/**
 * Get arguments from command line
 * @returns {Promise<{linkFile: *, outputFile: ({describe: string}|*|undefined), outputPath: ({describe: string}|*|undefined), width: undefined, url, height: undefined}>}
 */
async function getArguments() {

    const argv = yargs
        .help('h')
        .alias('h', 'help')
        .example('$0 -f links.txt', 'Use --file option to specify path to file')
        .example('$0 --width=800 --height=400', 'set the width and height of the viewport')
        .example('$0 --url=https://example.com', 'take a screenshot of a single URL (ignores link file)')
        .example('$0 --outputPath=screenshots', 'save screenshots to a custom folder')
        .example('$0 --outputFile=screenshots.jpg', 'save screenshots to a single file')
        .options({
            'file': {
                alias: 'f',
                describe: 'Path to link file',
                type: 'string'
            }
        })
        //add a new option for "width":
        .options({
            'width': {
                describe: 'Width of the viewport',
                type: 'string'
            }
        })
        //add a new option for "height":
        .options({
            'height': {
                describe: 'Height of the viewport',
                type: 'string'
            }
        })
        //add output path option:
        .options({
            'outputPath': {
                describe: 'Path to save screenshots',
            }
        })
        .options({
            'outputFile': {
                describe: 'Save screenshots to a single file',
            }
        })
        //add a new option for "url":
        .options({
            'url': {
                describe: 'URL to screenshot',
                type: 'string'
            }
        }).argv;

    return {
        linkFile: argv.file,
        url: argv.url,
        width: argv.width || undefined,
        height: argv.height || undefined,
        outputPath: argv.outputPath || undefined,
        outputFile: argv.outputFile || undefined
    }
}

/**
 * Read config file
 * @returns {Promise<{linkFile: string, tasks: [{browser: string, name: string, width: number, enabled: boolean, height: number}]}|any>}
 */
async function readConfigFile() {
    try {
        const data = await fs.readFile(confFile, 'utf-8');
        return JSON.parse(data);
    } catch (err) {
        console.log(`Config file ${confFile} does not exist. Using default config.`);

        const defaultConf = {
            "linkFile": "links.txt",
            "tasks": [
                {
                    "name": "screenshot",
                    "enabled": true,
                    "browser": "chrome",
                    "width": 1920,
                    "height": 1080
                }
            ]
        }

        return defaultConf;
    }
}

/**
 * Create folder if it does not exist
 * @param path
 * @returns {Promise<void>}
 */
async function createFolder(path) {
    try {
        await fs.access(path);
    } catch {
        await fs.mkdir(path, {recursive: true});
    }
}

/**
 * Take screenshots of a given task
 * @param task - object with task configuration
 * @param conf - object with global configuration
 * @param arguments - arguments passed from command line
 * @returns {Promise<void>}
 */
async function takeScreenshots(task, conf, arguments) {
    const currentScreenshotPath = arguments.outputPath ?? `${screenshotPath}/${task.name}_${task.browser}`;
    await createFolder(currentScreenshotPath);

    const browser = await puppeteer.launch({
        args: ['--no-sandbox', '--disable-setuid-sandbox'],
        headless: "new",
        product: task.browser,
        ignoreHTTPSErrors: true
    });

    const page = await browser.newPage();

    let height = arguments.height ?? task.height ?? 'full';
    let fullPage = false;
    if(height === 'full'){
        fullPage = true;
        height = 800;
    }

    const viewPortCongig = {
        width: parseInt(arguments.width ?? task.width ?? 1920),
        height: parseInt(height),
        deviceScaleFactor: 1
    };
    await page.setViewport(viewPortCongig);

    let links = [];
    const currentLinkFile = arguments.linkFile ?? task.linkFile ?? conf.linkFile ?? "links.txt";


    if (arguments.url) {
        links = [arguments.url];
    } else {
        links = (await fs.readFile(currentLinkFile, 'utf-8')).split('\n');
    }

    try {
        for (const link of links) {
            if (!link) continue;

            const baseUrl = task.baseUrl ?? '';
            const finalUrl = link.startsWith('http') ? link : `${baseUrl}${link}`;

            console.log(`Taking screenshot of ${finalUrl}...`);
            await page.goto(finalUrl, {waitUntil: 'networkidle0'});

            const outputFile = arguments.outputFile ?? `${link.replace(/[^a-z0-9]/gi, '-')}.jpg`;
            const screenshotParams = {
                path: `${currentScreenshotPath}/${outputFile}`,
                fullPage: fullPage
            };
            await page.screenshot(screenshotParams);
        }

        await browser.close();

    } catch (err) {

        if (err.code === 'ENOENT') {
            console.log(`Link file ${currentLinkFile} does not exist.`);
        }

    }
}


/**
 * Main function
 */
(async () => {
    const conf = await readConfigFile();
    const arguments = await getArguments();
    for (const task of conf.tasks) {
        if (!task.enabled ?? true) continue;
        await takeScreenshots(task, conf, arguments);
    }

    console.log("Done!");
    process.exit(0);
})();
